src/platform/packages/shared/kbn-data-grid-in-table-search/src/use_data_grid_in_table_search.tsx

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import React, { useMemo, useRef, useState } from 'react';
+import type { SerializedStyles } from '@emotion/react';
+import type { EuiDataGridProps, EuiDataGridRefProps } from '@elastic/eui';
+import { InTableSearchControl, InTableSearchControlProps } from './in_table_search_control';
+import { RenderCellValueWrapper } from './types';
+import { wrapRenderCellValueWithInTableSearchSupport } from './wrap_render_cell_value';
+
+export interface UseDataGridInTableSearchProps
+  extends Pick<InTableSearchControlProps, 'rows' | 'visibleColumns'> {
+  enableInTableSearch?: boolean;
+  dataGridWrapper: HTMLElement | null;
+  dataGridRef: React.RefObject<EuiDataGridRefProps | null>;
+  cellContext: EuiDataGridProps['cellContext'] | undefined;
+  pagination: EuiDataGridProps['pagination'] | undefined;
+  renderCellValue: EuiDataGridProps['renderCellValue'];
+}
+
+export interface UseDataGridInTableSearchState {
+  inTableSearchTerm: string;
+  inTableSearchTermCss?: SerializedStyles;
+}
+
+export interface UseDataGridInTableSearchReturn {
+  inTableSearchTermCss?: UseDataGridInTableSearchState['inTableSearchTermCss'];
+  inTableSearchControl: React.JSX.Element | undefined;
+  cellContextWithInTableSearchSupport: EuiDataGridProps['cellContext'];
+  renderCellValueWithInTableSearchSupport: RenderCellValueWrapper;
+}
+
+export const useDataGridInTableSearch = (
+  props: UseDataGridInTableSearchProps
+): UseDataGridInTableSearchReturn => {
+  const {
+    enableInTableSearch = true,
+    dataGridWrapper,
+    dataGridRef,
+    visibleColumns,
+    rows,
+    renderCellValue,
+    pagination,
+    cellContext,
+  } = props;
+  const isPaginationEnabled = Boolean(pagination);
+  const pageSize = (isPaginationEnabled && pagination?.pageSize) || null;
+  const onChangePage = pagination?.onChangePage;
+  const pageIndexRef = useRef<number>();
+  pageIndexRef.current = pagination?.pageIndex ?? 0;
+
+  const renderCellValueWithInTableSearchSupport = useMemo(
+    () => wrapRenderCellValueWithInTableSearchSupport(renderCellValue),
+    [renderCellValue]
+  );
+
+  const [{ inTableSearchTerm, inTableSearchTermCss }, setInTableSearchState] =
+    useState<UseDataGridInTableSearchState>(() => ({ inTableSearchTerm: '' }));
+
+  const inTableSearchControl = useMemo(() => {
+    if (!enableInTableSearch) {
+      return undefined;
+    }
+    const controlsCount = dataGridWrapper
+      ? dataGridWrapper.querySelectorAll('.euiDataGridHeaderCell--controlColumn').length
+      : 0;
```

Comment:
Would also be nice to have a cleaner way of knowing this, but again a minor concern.

Reply:
Looks like I was in the mood of using `querySelector` in many places here. So old school 😀